feat: Add pipeline run parallelism config

sduvvuri1603 opened the pull request:
## Summary
- Follow-up to the cleanup of unused mutex_name/semaphore_key: adds pipeline_run_parallelism as the official workflow-level concurrency knob so pipeline versions can declare how many runs may execute concurrently.
- The feature Pipeline_run_parallelism lets a pipeline version declare how many runs are allowed to execute at the same time, using Argo semaphores backed by a ConfigMap for workflow-level concurrency 
- Threads pipeline_run_parallelism through SDK → IR → compiler → backend; compiler validates (>0) without setting spec.parallelism, and backend CreateRun upserts the kfp-pipeline-config ConfigMap keyed by PipelineVersionId and wires Argo spec.synchronization.semaphores via configMapKeyRef.


## Validation
- SDK and backend goldens now include the updated sample, showing consistent IR and Argo outputs with the parallelism limit.
- Built custom API server and driver images from this branch, loaded them into a kind cluster, ran the sample, and confirmed that the number of simultaneously running component pods never exceeded the configured limit.
- Added the parallelism validation helper to the e2e suite (`e2e_utils.go` + invocation in `pipeline_e2e_test.go`), rebuilt the test cluster with the fresh backend images, exercised the focused pipeline_run_parallelism scenario, and then ran the end-to-end suite to confirm the new check passes with the concurrency cap enforced.

Note:
Hardened .github/actions/test-and-report/action.yml: the workflow now waits for the ml-pipeline deployment, port-forwards svc/ml-pipeline to localhost:8888, and curls /apis/v2beta1/healthz (TLS-aware). This eliminates the dial tcp [::1]:8888: connect: connection refused failures that intermittently broke the parallelism/recurring-run E2E suite.

Follow up to [PR](https://github.com/kubeflow/pipelines/pull/12401) - remove unused semaphore_key and mutex_name fields

hbelmiro commented:
/retest

nsingla reviewed (changes requested):
I think we need a pipeline in `test_data/valid/*compiled/essential` that overrides this value.

nsingla commented on test_data/sdk_compiled_pipelines/valid/critical/pipeline_with_workspace.py:
```diff
@@ -53,6 +53,7 @@ def read_from_workspace(file_path: str) -> str:
     name="pipeline-with-workspace",
     description="A pipeline that demonstrates workspace functionality",
     pipeline_config=dsl.PipelineConfig(
+        pipeline_run_parallelism=3,
```
can we actually set this to None here if we have an explicit test to test +ve values?

nsingla commented on test_data/sdk_compiled_pipelines/valid/essential/pipeline_with_run_parallelism.py:
```diff
@@ -0,0 +1,36 @@
+
+@dsl.pipeline(
+    name='pipeline-with-run-parallelism',
+    pipeline_config=dsl.PipelineConfig(pipeline_run_parallelism=7),
```
Isn't 7 too high when the number of tasks in this pipeline is just 1? May be you should add more components to it or add a parallelFor loop and iterate over > pipeline_run_parallelism constants, so that we can validate that the config actually works.
Also what validation logic did you add to confirm the number of tasks created for a pipeline with this config?

nsingla reviewed (approved):
/lgtm

sduvvuri1603 commented:
/retest

hbelmiro reviewed (commented):
@sduvvuri1603 I'm not sure about what exactly this PR is supposed to do (probably because I don't have any context on this parallelism work that has been done).
The PR's title is a bit vague. It's ok the title to be vague since you can elaborate in the description. But in the description you explain how things were done but don't explain what the new config is supposed to do.

You do explain though in `README.md`:

> New in master: `dsl.PipelineConfig` now accepts an optional `pipeline_run_parallelism` integer to cap concurrent task execution for a run. The backend stores the requested limit in a shared ConfigMap and surfaces it to Argo Workflows via `spec.parallelism`.

So, I'm thinking this PR is supposed to add only concurrency task execution for a run, not cross-workflow semaphores.

But you also mention and created a ConfigMap `const pipelineParallelismConfigMapName = "kfp-pipeline-config"`, which is for cross-workflow semaphores. That ConfigMap is never used though.

This is confusing me. Can you clarify that before I dive into the code?

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -50,6 +50,8 @@ import (
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 )
 
+const pipelineParallelismConfigMapName = "kfp-pipeline-config"
```
You're correct: the ConfigMap is created but not read. The parallelism limit is enforced by Argo's spec.parallelism, which is set directly during compilation (not from the ConfigMap). The ConfigMap stores values keyed by PipelineVersionId for audit/persistence only. This feature is for task-level concurrency within a single run, not cross-workflow semaphores, so spec.synchronization.semaphore.configMapKeyRef is not needed.

sduvvuri1603 replied:
changing logic to make sure argo reads from configmap and I use the code reference of "Using a semaphore configured by a ConfigMap"

sduvvuri1603 commented on proposals/11875-pipeline-workspace/README.md:
```diff
@@ -394,8 +394,12 @@ index d986a048b..866696419 100644
+  // Maximum number of tasks that can be scheduled simultaneously for a single
+  // pipeline run.
+  optional int32 pipeline_run_parallelism = 3;
  }
```
You're right—sorry about that. This was added when documenting the proto changes, but it doesn't belong here since this KEP is about workspace, not parallelism. I'll revert back my changes!

sduvvuri1603 commented:
/retest

hbelmiro commented:
Thank you for the clarification @sduvvuri1603.

> ... for persistence/audit purposes only

Is it really a requirement? The parallelism is already persisted in the pipeline spec (which is stored when the version is created). What scenario requires reading from this ConfigMap rather than the pipeline spec?

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

hbelmiro commented on backend/src/apiserver/resource/resource_manager_test.go:
```diff
@@ -1936,6 +1936,72 @@ func TestCreateRun_ThroughPipelineVersion(t *testing.T) {
+		"kubeflow",
+	)
+	version, err := manager.CreatePipelineVersion(pv)
+	assert.NoError(t, err)
```
```suggestion
	require.NoError(t, err)
```

hbelmiro commented on backend/src/apiserver/resource/resource_manager_test.go:
```diff
@@ -1936,6 +1936,72 @@ func TestCreateRun_ThroughPipelineVersion(t *testing.T) {
+
+	createdRun, err := manager.CreateRun(context.Background(), run)
+	assert.NoError(t, err)
+	assert.NotNil(t, createdRun)
```
```suggestion
	require.NoError(t, err)
	require.NotNil(t, createdRun)
```

hbelmiro commented on backend/src/apiserver/resource/resource_manager_test.go:
```diff
@@ -1936,6 +1936,72 @@ func TestCreateRun_ThroughPipelineVersion(t *testing.T) {
+
+	configMap, err := store.k8sCoreClientFake.ConfigMapClient("kubeflow").Get(
+		context.Background(), pipelineParallelismConfigMapName, v1.GetOptions{})
+	assert.NoError(t, err)
```
```suggestion
	require.NoError(t, err)
```

hbelmiro commented on backend/src/apiserver/resource/resource_manager_test.go:
```diff
@@ -1936,6 +1936,72 @@ func TestCreateRun_ThroughPipelineVersion(t *testing.T) {
+			assert.Equal(t, pipelineParallelismConfigMapName, semaphore.ConfigMapKeyRef.Name)
+			assert.Equal(t, version.UUID, semaphore.ConfigMapKeyRef.Key)
+		}
+	}
```
```suggestion
    require.NotNil(t, workflow.Spec.Synchronization)
    require.NotNil(t, workflow.Spec.Synchronization.Semaphores)
    require.Len(t, workflow.Spec.Synchronization.Semaphores, 1)
    semaphore := workflow.Spec.Synchronization.Semaphores[0]
    assert.NotNil(t, semaphore.ConfigMapKeyRef)
    assert.Equal(t, pipelineParallelismConfigMapName, semaphore.ConfigMapKeyRef.Name)
    assert.Equal(t, version.UUID, semaphore.ConfigMapKeyRef.Key)
```

hbelmiro commented on backend/src/cache/client/kubernetes_core_fake.go:
```diff
@@ -34,22 +36,39 @@ func (c *FakeKuberneteCoreClient) PodClient(namespace string) v1.PodInterface {
+
 func NewFakeKuberneteCoresClient() *FakeKuberneteCoreClient {
-	return &FakeKuberneteCoreClient{&FakePodClient{}}
+	clientset := kubernetesfake.NewSimpleClientset()
```
`NewSimpleClientset` is deprecated.

hbelmiro commented on backend/src/cache/client/kubernetes_core_fake.go:
```diff
@@ -34,22 +36,39 @@ func (c *FakeKuberneteCoreClient) PodClient(namespace string) v1.PodInterface {
+		podClientFake: &FakeBadPodClient{},
+		coreClient:    clientset.CoreV1(),
+	}
 }
```
This function is not used and can be removed.

sduvvuri1603 replied:
removed unused function

hbelmiro commented on test_data/sdk_compiled_pipelines/valid/critical/pipeline_with_workspace.py:
```diff
@@ -53,6 +53,7 @@ def read_from_workspace(file_path: str) -> str:
     name="pipeline-with-workspace",
     description="A pipeline that demonstrates workspace functionality",
     pipeline_config=dsl.PipelineConfig(
+        pipeline_run_parallelism=None,
```
Even this working as a test for no parallelism, I think we should create a specific pipeline to test this particular scenario. My concern is that this pipeline may change or be removed in the future for whatever reason and we may not notice that this also tests no parallelism (beside workspace). We would end up with no tests for no parallelism.

hbelmiro commented on backend/test/end2end/utils/e2e_utils.go:
```diff
@@ -93,6 +96,166 @@ func ValidateComponentStatuses(runClient *apiserver.RunClient, k8Client *kuberne
+		gomega.Expect(running).To(gomega.BeNumerically(">=", 0))
+		gomega.Expect(running).To(gomega.BeNumerically("<=", limit))
+	}
+}
```
`ValidateParallelismIfConfigured` appears to be dead/broken code and has some problems:

1. It never executes: The only pipeline with `pipelineRunParallelism` set is `pipeline_with_run_parallelism`, but that's explicitly skipped at line 110.

2. Wrong abstraction: The function validates task-level concurrency (concurrent tasks within a single run)

3. Would fail if enabled: If the skip at line 110 were removed, the test would fail because `pipeline_with_run_parallelism` has 5 ParallelFor tasks, but `pipelineRunParallelism=2`.

Suggestion: Remove this function entirely. Workflow-level concurrency is already tested by `ValidateWorkflowParallelismAcrossRuns`.

sduvvuri1603 replied:
Thanks for pointing that out!
Removed redundant and unused function

hbelmiro commented on backend/src/v2/compiler/argocompiler/argo.go:
```diff
@@ -150,6 +150,19 @@ func Compile(jobArg *pipelinespec.PipelineJob, kubernetesSpecArg *pipelinespec.S
+			}
+			// Note: We don't set wf.Spec.Parallelism here because that limits tasks within a single workflow.
+			// Instead, we use synchronization.semaphores with ConfigMap to limit concurrent workflow runs.
+			// The synchronization is configured in resource_manager.go after the workflow is created.
```
This comment is not necessary.

sduvvuri1603 replied:
Removed

hbelmiro commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,88 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+	// Get the parallelism value
+	value := pipelineConfig.GetPipelineRunParallelism()
+	if value <= 0 {
+		return 0, false
```
We should return an error here.

hbelmiro commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,88 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+
+func (r *ResourceManager) upsertPipelineParallelismConfigMap(ctx context.Context, namespace, key string, parallelism int64) error {
+	if key == "" {
+		return nil
```
`key` being nil is a bug. We should not hide it. Return an error instead.

sduvvuri1603 replied:
modified code to return error "key cannot be empty"  when key is nil in both 'upsertPipelineParallelismConfigMap' and 'deletePipelineParallelismConfigMapEntry'.

hbelmiro commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -562,6 +647,38 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
+			ctx, k8sNamespace, run.PipelineVersionId, parallelism); err != nil {
+			if apierrors.IsForbidden(errors.Cause(err)) {
+				glog.Warningf("Skipping pipeline_run_parallelism ConfigMap update in namespace %q: %v. Semaphore will not be configured.", k8sNamespace, err)
+				// Don't configure semaphores if ConfigMap creation fails
```
I think we should return an error.

hbelmiro commented on backend/src/cache/deployer/deploy-cache-service.sh:
```diff
@@ -34,8 +34,15 @@ mkdir -p "$HOME/bin"
+        # Can be overridden via KUBECTL_VERSION_FALLBACK env var.
+        stable_build_version="${KUBECTL_VERSION_FALLBACK:-v1.34.0}"
+    fi
+    kubectl_url="https://dl.k8s.io/release/${stable_build_version}/bin/linux/amd64/kubectl"
```
This change seems unrelated.

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager_test.go:
```diff
@@ -1936,6 +1936,72 @@ func TestCreateRun_ThroughPipelineVersion(t *testing.T) {
+	initEnvVars()
+
+	store := NewFakeClientManagerOrFatal(util.NewFakeTimeForEpoch())
+	defer store.Close()
```
Updated the defer statement to handle the error 

Added an  If statement where in  "if Close() returns an error, the test will report it." following Go pattern for handling errors in deferred cleanup calls.

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,88 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+			return util.Wrap(err, "failed to update pipeline parallelism ConfigMap entry")
+		}
+		return nil
+	}
```
Thanks a lot for that. Added context cancellation check at the start of the retry loop to avoid infinite retries.

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,88 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+	return int64(value), true
+}
+
+func (r *ResourceManager) upsertPipelineParallelismConfigMap(ctx context.Context, namespace, key string, parallelism int64) error {
```
Added deletePipelineParallelismConfigMapEntry for removing a pipeline version's parallelism entry from the ConfigMap. This is a helper function for cleanup 
I added a helper function deletePipelineParallelismConfigMapEntry() for cleanup(removing pipeline version from configmap) when pipeline versions are deleted.

But it isn't called automatically yet due to a technical limitation I am facing:

ConfigMaps are namespace-scoped, and pipeline versions don't store namespace info.. When a pipeline version is deleted, we don't know which namespaces contain ConfigMap entries for that version, so how can I automatically clean them up? 

I searched up few potential cleanups:

1. A background job that periodically removes stale entries
2. Cleanup during run deletion when namespace is known
3. Manual cleanup operation

Should ConfigMap cleanup be implemented as part of this PR, or would it be better to handle it in a follow-up? If we should include it here, I can implement one of the approaches above (most likely cleanup during run deletion when we have namespace context).

Please let me know. Thanks :)

hbelmiro replied:
I think we can accept this as a known limitation and track cleanup in a follow-up issue with a proper discussion about the solution.

Thoughts @mprahl @HumairAK?

sduvvuri1603 commented on sdk/python/kfp/dsl/pipeline_config.py:
```diff
@@ -96,5 +96,25 @@ def set_kubernetes_config(self,
+        return self._pipeline_run_parallelism
+
+    @pipeline_run_parallelism.setter
+    def pipeline_run_parallelism(self, value: Optional[int]) -> None:  # pylint: disable=attribute-defined-outside-init
```
Updated code with suggested change to avoid pylint disable

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,88 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
 
+// extractPipelineRunParallelism extracts pipeline_run_parallelism from the template's platform spec.
+// It navigates: template -> platformSpec -> platforms["kubernetes"] -> pipelineConfig -> pipelineRunParallelism
+func extractPipelineRunParallelism(tmpl template.Template) (int64, bool) {
```
Thanks for the feedback! I've updated the function to return `(int64, error)` instead of `(int64, bool)`. 

The function now:
- Returns `(0, nil)` when parallelism is not configured (normal case)
- Returns `(0, error)` when validation fails (e.g., value <= 0)
- Returns `(value, nil)` when successfully extracted

The caller now explicitly checks for errors and returns them, preventing validation failures from being silently ignored.

sduvvuri1603 commented on sdk/python/kfp/compiler/compiler_test.py:
```diff
@@ -1153,7 +1153,8 @@ def my_pipeline(input1: str):
                 pipeline_config=dsl.PipelineConfig(
-                    workspace=dsl.WorkspaceConfig(size='25Gi'),),
+                    workspace=dsl.WorkspaceConfig(size='25Gi'),
+                    pipeline_run_parallelism=5),
```
Removed irrelevant code 'pipeline_run_parallelism=5'

sduvvuri1603 commented on sdk/python/kfp/dsl/pipeline_config.py:
```diff
@@ -96,5 +96,25 @@ def set_kubernetes_config(self,
+                    'pipeline_run_parallelism must be a positive integer.')
+            self._pipeline_run_parallelism = value
+        else:
+            self._pipeline_run_parallelism = None
```
That is actually a good idea and will clearly indicate the intention of the feature. Changing the name to "pipeline_version_concurrency_limit"

mprahl replied:
How about `max_active_runs`? It's implied it's specific to the pipeline version since the PipelineConfig is always tied to the version.

sduvvuri1603 commented:
/retest

hbelmiro commented on manifests/kustomize/base/pipeline/ml-pipeline-apiserver-role.yaml:
```diff
@@ -14,6 +14,16 @@ rules:
+  - list
+  - create
+  - update
+  - patch
```
This Role grants ConfigMap permissions on ALL configmaps in the namespace. The ClusterRole uses `resourceNames: [kfp-pipeline-config]` to restrict operations to only the specific ConfigMap we manage.

Can we apply the same restriction here for least-privilege?

```suggestion
- apiGroups:
  - ""
  resources:
  - configmaps
  verbs:
  - create
- apiGroups:
  - ""
  resourceNames:
  - kfp-pipeline-config
  resources:
  - configmaps
  verbs:
  - get
  - update
  - patch
```

Note: `list` doesn't appear to be used (upsert calls Get, not List).

hbelmiro commented on sdk/python/kfp/dsl/pipeline_config.py:
```diff
@@ -96,5 +96,28 @@ def set_kubernetes_config(self,
+            if not isinstance(value, int):
+                raise ValueError(
+                    'pipeline_version_concurrency_limit must be an integer if specified.'
+                )
```
This check is unreachable. The type hint declares `value: Optional[int]`, and after the `if value is not None` check, the type is narrowed to `int`. `isinstance(value, int)` is always True.

hbelmiro commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,131 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
 
+// extractPipelineVersionConcurrencyLimit extracts pipeline_version_concurrency_limit from the template's platform spec.
+// It navigates: template -> platformSpec -> platforms["kubernetes"] -> pipelineConfig -> pipelineVersionConcurrencyLimit
+func extractPipelineVersionConcurrencyLimit(tmpl template.Template) (int64, error) {
```
This can return `int`.

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -562,6 +690,43 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
+			// In multi-user mode, ConfigMap creation may fail due to permissions.
+			// Log a warning and continue without the concurrency limit rather than failing the run.
+			if apierrors.IsForbidden(err) {
+				glog.Warningf("Failed to create pipeline_version_concurrency_limit ConfigMap in namespace %s due to insufficient permissions. Run will proceed without concurrency limit: %v", k8sNamespace, err)
```
That makes sense. Made changes to fail the run

sduvvuri1603 commented on test_data/sdk_compiled_pipelines/valid/critical/pipeline_with_workspace.py:
```diff
@@ -53,6 +53,7 @@ def read_from_workspace(file_path: str) -> str:
     name="pipeline-with-workspace",
     description="A pipeline that demonstrates workspace functionality",
     pipeline_config=dsl.PipelineConfig(
+        pipeline_version_concurrency_limit=None,
```
Removed

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -562,6 +690,43 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
+							Key: run.PipelineVersionId,
+						},
+					},
+				}
```
Used append to avoid overriding for future extensibility.

sduvvuri1603 commented on backend/src/apiserver/client/kubernetes_core_fake.go:
```diff
@@ -35,13 +38,21 @@ func (c *FakeKuberneteCoreClient) PodClient(namespace string) v1.PodInterface {
+	}
 }
 
 type FakeKubernetesCoreClientWithBadPodClient struct {
```
Made changes to add coreClient field and Reuse it.

hbelmiro reviewed (commented):
The ConfigMap cleanup flow is untested. Testing this properly would require simulating runs across multiple namespaces and verifying cleanup in each, which is complex. Given that the cleanup is best-effort (logs warnings but doesn't fail deletion), this is acceptable for now, but worth noting as a gap.

I left one more final comment. Once addressed, I think we're fine.

sduvvuri1603 commented on backend/src/apiserver/storage/run_store.go:
```diff
@@ -793,3 +798,45 @@ func (s *RunStore) scanRowsToRunMetrics(rows *sql.Rows) ([]*model.RunMetric, err
+	if err != nil {
+		return nil, util.NewInternalServerError(err, "Failed to query distinct namespaces for pipeline version %s: %v", pipelineVersionID, err)
+	}
+	defer rows.Close()
```
Came up with this  idea, pls verify if it's correct:

Handled the error from rows.Close() using named return values and a defer closure.

Implementation:
- Function uses named return (namespaces []string, err error)
- Defer closure checks rows.Close() error: if closeErr := rows.Close(); closeErr != nil && err == nil
- Only sets the close error if no other error occurred (preserves scan/query errors)
- Used scanErr and rowErr to avoid shadowing the named return err

sduvvuri1603 replied:
Yep, will revert back changes once CI runs complete

hbelmiro replied:
Oh, you've actually handled them. Good.

hbelmiro reviewed (commented):
/lgtm

mprahl commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,131 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+	if !ok || v2Spec == nil {
+		return 0, nil
+	}
+	// Get the platform spec (already public method)
```
Remove any self-evidence comments please

mprahl commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -562,6 +690,41 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
+			ctx, k8sNamespace, run.PipelineVersionId, int(maxActiveRuns)); err != nil {
+			// In multi-user mode, provide a more helpful error message if it's a permission issue
+			if common.IsMultiUserMode() && apierrors.IsForbidden(err) {
+				return nil, util.Wrapf(err, "Failed to persist max_active_runs configuration. The API server needs ConfigMap create/update permissions in namespace %s. Please ensure the ml-pipeline service account has the necessary RBAC permissions", k8sNamespace)
```
You already updated the deployment manifests. No need for this special case error message.

Just log any encountered error and let the admin sort it out of they are not using updated manifests.

sduvvuri1603 replied:
Updated to log error instead of failing

mprahl commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -1829,6 +1992,26 @@ func (r *ResourceManager) DeletePipelineVersion(pipelineVersionId string) error
+	} else {
+		ctx := context.Background()
+		for _, namespace := range namespaces {
+			if err := r.deletePipelineParallelismConfigMapEntry(ctx, namespace, pipelineVersionId); err != nil {
```
This really should be abstracted as a pipeline engine agnostic way. Perhaps you can add a new method on `ExecutionClient` that `WorkflowClient` implements? A method name could be `OnDeletePipelineVersion` or something.

mprahl commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -50,6 +51,8 @@ import (
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 )
 
+const pipelineParallelismConfigMapName = "kfp-pipeline-config"
```
This ConfigMap is very Argo Workflows specific and only for semaphores, so let's consider naming it like that. The current name sounds like it's a generic KFP config.

sduvvuri1603 replied:
That totally makes sense. 
Updated name to kfp-argo-workflow-semaphores

sduvvuri1603 commented on api/v2alpha1/pipeline_spec.proto:
```diff
@@ -1176,4 +1176,9 @@ message PipelineConfig {
   optional WorkspaceConfig workspace = 2;
+
+  // Maximum number of concurrent runs allowed for this pipeline version.
+  // When set, Argo Workflows will use synchronization semaphores to limit
```
Updated comment to shortly describe behavior: "When set, limits the number of simultaneously active runs for the same pipeline version."

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,131 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+			return util.Wrap(err, "failed to retrieve pipeline parallelism ConfigMap for cleanup")
+		}
+
+		if current.Data == nil {
```
Removed the check since nil map is allowed in Go

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -153,6 +156,131 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+		}
+
+		delete(current.Data, key)
+		if _, err := configMaps.Update(ctx, current, v1.UpdateOptions{}); apierrors.IsConflict(err) {
```
- Switched from Get + Update to a single PATCH call
- Replaced Get + Create/Update with server-side apply using ConfigMapApplyConfiguration

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -562,6 +690,41 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
 	executionSpec.SetExecutionNamespace(k8sNamespace)
 
+	// Extract max active runs from template and configure synchronization semaphores
+	maxActiveRuns, err := extractMaxActiveRuns(tmpl)
```
Yes, recurring runs are supported.
The runs go through the same CreateRun path (lines 615-692) before setting the recurring run owner reference so they respect the max_active_runs limit

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -562,6 +690,41 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
+	}
+	if maxActiveRuns > 0 && run.PipelineVersionId != "" {
+		// Create/update ConfigMap with semaphore limit
+		if err := r.upsertPipelineParallelismConfigMap(
```
Shifted Argo-specific logic into WorkflowInterface.Create() where  Argo Workflow is submitted

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -562,6 +690,41 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
+			wf.Spec.Synchronization.Semaphores = append(
+				wf.Spec.Synchronization.Semaphores,
+				&workflowapi.SemaphoreRef{
+					ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
```
Thanks for that!
Moved the semaphore config to (`argocompiler.Compile()`). The semaphores are now configured during compilation using `PipelineConfig.MaxActiveRuns` and `PipelineVersionId` from the compiler options. 

ConfigMap upsert remains in `WorkflowInterface.Create()` for runtime operations.

sduvvuri1603 commented on backend/src/apiserver/storage/run_store.go:
```diff
@@ -95,6 +95,11 @@ type RunStoreInterface interface {
+	// Gets distinct namespaces where runs for a pipeline version exist.
+	// This is used to identify all namespaces that may contain ConfigMap entries
+	// for a pipeline version when cleaning up resources.
+	GetDistinctNamespacesForPipelineVersion(pipelineVersionID string) ([]string, error)
```
Updated method name

sduvvuri1603 commented on backend/src/cache/client/kubernetes_core.go:
```diff
@@ -23,6 +24,10 @@ func (c *KubernetesCore) PodClient(namespace string) v1.PodInterface {
 }
 
+func (c *KubernetesCore) ConfigMapClient(namespace string) v1.ConfigMapInterface {
+	return c.coreV1Client.ConfigMaps(namespace)
```
I just searched it up and found that informer cache is a good optimization for read operations. 

Since we're only doing write operations rn, should I add it now or as a follow up PR for future implementation?

sduvvuri1603 commented on backend/test/end2end/pipeline_e2e_test.go:
```diff
@@ -266,5 +266,8 @@ func validatePipelineRunSuccess(pipelineFile string, pipelineDir string, testCon
 	}
 	compiledWorkflow := workflowutils.UnmarshallWorkflowYAML(filepath.Join(testutil.GetCompiledWorkflowsFilesDir(), pipelineFile))
 	e2e_utils.ValidateComponentStatuses(runClient, k8Client, testContext, createdRunID, compiledWorkflow)
+	if limit, ok := e2e_utils.MaxActiveRuns(pipelineFilePath); ok {
```
Thanks for the feedback! I've added all four test scenarios you requested in `backend/test/end2end/pipeline_e2e_test.go`:

1. Test 1: Single pipeline version with > MaxParallelism runs - verifies only MaxParallelism runs are active
2. Test 2: Same pipeline with different versions - verifies all runs from different versions can be active independently
3. Test 3: Mixed scenario (same version + different versions) - verifies limits are enforced per version even when mixed
4. Test 4: Different pipelines - verifies runs from different pipelines don't interfere with each other

All tests use the `pipeline_with_max_active_runs.yaml` pipeline and validate that parallelism limits are correctly enforced per pipeline version. The helper function `ValidateParallelismAcrossRuns` tracks active runs per version and validates limits independently.

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

hbelmiro commented on backend/src/common/util/workflow.go:
```diff
@@ -865,24 +875,129 @@ func (wc *WorkflowClient) Compare(old, new interface{}) bool {
+				glog.Warningf("Failed to create Kubernetes clientset for ConfigMap cleanup: %v", err)
+				return
+			}
+			ctx := context.Background()
```
Using `context.Background()` for ConfigMap cleanup operations means these operations cannot be canceled even if the parent operation is canceled or times out. This could lead to orphaned goroutines continuing to run after the parent context is done. You can pass a derived context with a timeout instead:

```suggestion
			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
			defer cancel()
```

sduvvuri1603 replied:
Updated code to include timeout

sduvvuri1603 commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -1829,6 +1886,19 @@ func (r *ResourceManager) DeletePipelineVersion(pipelineVersionId string) error
+	// Clean up engine-specific resources for this pipeline version asynchronously.
+	// This identifies all namespaces where runs of this pipeline version exist and delegates cleanup to the execution client.
+	defer func() {
+		go func() {
```
Yes, this is intentional. 

Based on Matt's review I used defer func() { go func() { to ensure the namespace lookup runs even if the function returns early (via defer) while remaining asynchronous and non-blocking (via go func()). 

This prevents the database query from blocking the HTTP response. The pattern then delegates to OnDeletePipelineVersion which handles the actual ConfigMap cleanup asynchronously.

hbelmiro reviewed (changes requested):
The constant `pipelineParallelismConfigMapName = "kfp-argo-workflow-semaphores"` is defined in 5 different files:

- `backend/src/v2/compiler/argocompiler/argo.go:155`
- `backend/src/crd/controller/scheduledworkflow/controller.go:617`
- `backend/src/common/util/workflow.go:940`
- `backend/src/apiserver/resource/resource_manager.go:54`
- `backend/src/apiserver/client/workflow_fake.go:42`

Can you please move this constant to a shared location (e.g., `backend/src/common/util/consts.go`) and import it from there?

sduvvuri1603 commented on backend/src/crd/controller/scheduledworkflow/controller.go:
```diff
@@ -588,6 +603,82 @@ func (c *Controller) submitNewWorkflowIfNotAlreadySubmitted(
+// extractMaxActiveRunsFromWorkflow extracts max_active_runs from embedded workflow spec.
+// It extracts the pipeline version ID from semaphores and looks up max_active_runs
+// from the pipeline version API (since it's not stored in the compiled workflow).
+func (c *Controller) extractMaxActiveRunsFromWorkflow(ctx context.Context, workflow *commonutil.Workflow, fallbackPipelineVersionID, pipelineID string) (maxActiveRuns int32, pipelineVersionID string) {
```
I originally returned both the limit and the resolved pipeline version ID so the caller could annotate the workflow with both max_active_runs and pipeline_version_id. 

If that looks like juggling between 2 values, I’ve moved that annotation into the helper itself—it still sets AnnotationKeyPipelineVersionID on the workflow when it finds a value—so the caller only needs the numeric limit. The behavior is unchanged; the helper now returns (int32, error) as you suggested.

sduvvuri1603 commented on backend/src/crd/controller/scheduledworkflow/controller.go:
```diff
@@ -588,6 +603,82 @@ func (c *Controller) submitNewWorkflowIfNotAlreadySubmitted(
+// extractMaxActiveRunsFromWorkflow extracts max_active_runs from embedded workflow spec.
+// It extracts the pipeline version ID from semaphores and looks up max_active_runs
+// from the pipeline version API (since it's not stored in the compiled workflow).
+func (c *Controller) extractMaxActiveRunsFromWorkflow(ctx context.Context, workflow *commonutil.Workflow, fallbackPipelineVersionID, pipelineID string) (maxActiveRuns int32, pipelineVersionID string) {
```
I originally returned both the limit and the resolved pipeline version ID so the caller could annotate the workflow with both max_active_runs and pipeline_version_id. 

If that looks like juggling between 2 values, I’ve moved that annotation into the helper itself—it still sets AnnotationKeyPipelineVersionID on the workflow when it finds a value—so the caller only needs the numeric limit. The behavior is unchanged; the helper now returns (int32, error) as you suggested.

sduvvuri1603 commented on backend/src/crd/controller/scheduledworkflow/controller.go:
```diff
@@ -588,6 +603,82 @@ func (c *Controller) submitNewWorkflowIfNotAlreadySubmitted(
+// extractMaxActiveRunsFromWorkflow extracts max_active_runs from embedded workflow spec.
+// It extracts the pipeline version ID from semaphores and looks up max_active_runs
+// from the pipeline version API (since it's not stored in the compiled workflow).
+func (c *Controller) extractMaxActiveRunsFromWorkflow(ctx context.Context, workflow *commonutil.Workflow, fallbackPipelineVersionID, pipelineID string) (maxActiveRuns int32, pipelineVersionID string) {
```
Removed it based on my updated logic of moving the pipeline-version annotation into the helper, the parameter becomes redundant

sduvvuri1603 commented on backend/test/end2end/utils/e2e_utils.go:
```diff
@@ -93,6 +93,157 @@ func ValidateComponentStatuses(runClient *apiserver.RunClient, k8Client *kuberne
+		return 0, false
+	}
+	return val, true
+}
```
Sorry for overlooking on this. Adjusted logic to explicitly show errrors and return values

sduvvuri1603 commented on backend/test/end2end/utils/e2e_utils.go:
```diff
@@ -93,6 +93,157 @@ func ValidateComponentStatuses(runClient *apiserver.RunClient, k8Client *kuberne
+		for _, runInfo := range runInfos {
+			run := testutil.GetPipelineRun(runClient, &runInfo.RunID)
+			if run.State == nil {
+				activeByVersion[runInfo.PipelineVersionID]++
```
- Yes—when we call GetPipelineRun immediately after scheduling, the API often returns a Run where state hasn’t been populated yet. 
- That doesn’t mean the run is terminal; it’s just not initialized yet. So, If we don't treat nil as “active", semaphore checks would conclude there are zero active runs and stop validating. By counting nil as active, we keep waiting until the run reaches a real terminal state before asserting limits. 
- Once state becomes PENDING or RUNNING, it continues to be counted; once it’s SUCCEEDED/FAILED, it drops out.

 So keeping the run.State == nil branch is intentional.

sduvvuri1603 commented on backend/test/testutil/pipeline_recurring_run_utils.go:
```diff
@@ -43,11 +43,18 @@ func GetRecurringRun(client *api_server.RecurringRunClient, runID string) *recur
+	}
+	if client == nil {
+		logger.Log("Skipping deletion of recurring run id=%s because recurring run client is nil", runID)
+		return
```
If a test forgets to supply a client or ID, it now fails immediately instead of skipping the delete

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

sduvvuri1603 commented:
/retest

sduvvuri1603 commented on sdk/python/test/compilation/pipeline_compilation_test.py:
```diff
@@ -321,6 +323,13 @@ def __repr__(self) -> str:
+            TestData(
+                pipeline_name='pipeline-with-max-active-runs',
+                pipeline_func=pipeline_with_max_active_runs,
+                pipline_func_args=None,
```
Addressed the typo in: [correct typo pipline_func_args -> pipeline_func_args](https://github.com/kubeflow/pipelines/pull/12833)

sduvvuri1603 commented:
/retest

hbelmiro commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -169,10 +161,35 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
+	}
+	value, okValue, err := v2Spec.MaxActiveRuns()
+	if err != nil {
+		return 0, err
```
Return the error.

hbelmiro commented on backend/src/common/util/workflow.go:
```diff
@@ -865,9 +875,64 @@ func (wc *WorkflowClient) Compare(old, new interface{}) bool {
 }
 
+func (wc *WorkflowClient) OnDeletePipelineVersion(pipelineVersionID string, namespaces []string) {
+	if pipelineVersionID == "" || len(namespaces) == 0 {
```
Let's return an error when `pipelineVersionID == ""`.

hbelmiro commented on backend/test/end2end/pipeline_e2e_test.go:
```diff
@@ -247,6 +256,368 @@ var _ = Describe("Upload and Verify Pipeline Run >", Label(FullRegression), func
+			limit, err := e2e_utils.MaxActiveRuns(pipelineFilePath)
+			Expect(err).NotTo(HaveOccurred(), "Pipeline should have max_active_runs configured")
+			Expect(limit).To(BeNumerically(">", 0), "max_active_runs should be greater than 0")
+			Expect(limit).To(BeNumerically(">", 0), "max_active_runs should be greater than 0")
```
```suggestion
			Expect(limit).To(BeNumerically(">", 0), "max_active_runs should be greater than 0")
```

hbelmiro commented on backend/test/end2end/pipeline_e2e_test.go:
```diff
@@ -247,6 +256,368 @@ var _ = Describe("Upload and Verify Pipeline Run >", Label(FullRegression), func
+			}
+		}
+
+		collectRunInfos := func(recurringRunID string, fallbackPipelineID string, fallbackPipelineVersionID string, targetRuns int) []e2e_utils.RunInfo {
```
This 45-line closure captures `runClient`, `experimentID`, and `testContext` from the outer scope. Consider extracting it as a standalone function with explicit parameters.

The `fallbackPipelineID` / `fallbackPipelineVersionID` parameters paper over missing data instead of surfacing it as a test failure. If a run created by a recurring run doesn't have a `PipelineVersionReference`, that's a bug worth catching, not a gap to fill with empty strings.

sduvvuri1603 commented on backend/src/apiserver/template/v2_template.go:
```diff
@@ -85,6 +85,27 @@ func (t *V2Spec) PlatformSpec() *pipelinespec.PlatformSpec {
 
+// MaxActiveRuns returns the configured max_active_runs from the Kubernetes platform spec.
+// The second return value indicates whether the field was explicitly set.
+func (t *V2Spec) MaxActiveRuns() (int32, bool, error) {
```
Since max_active_runs is optional.

We need to know two things:
1) the number (if it exists)
2) whether it was set at all

If we only return (int32, error), then 0 could mean: “not set”, or “set to 0 (invalid)”.
The bool tells us “was it set?” so we can differentiate both cases.

sduvvuri1603 commented on backend/test/end2end/pipeline_e2e_test.go:
```diff
@@ -247,6 +256,368 @@ var _ = Describe("Upload and Verify Pipeline Run >", Label(FullRegression), func
+		var pipelineFile = "essential/pipeline_with_max_active_runs.yaml"
+		var pipelineDir = "valid"
+
+		It("Test 1: > MaxParallelism runs of a single pipeline version - only MaxParallelism runs should be active", func() {
```
Sure, Removed. 
Thanks!

sduvvuri1603 commented on backend/src/crd/controller/scheduledworkflow/controller.go:
```diff
@@ -644,6 +662,98 @@ func (c *Controller) submitNewWorkflowIfNotAlreadySubmitted(
+		}
+	}
+
+	pipelineVersion, err := c.pipelineClient.GetPipelineVersion(ctx, &api.GetPipelineVersionRequest{
```
Good point. 
I updated the ScheduledWorkflow path to read pipelines.kubeflow.org/max_active_runs directly from the compiled workflow annotations and avoid the per‑run API lookup. 

The compiler now stamps max_active_runs (and pipeline version ID when present) onto the workflow spec, so the controller can use it without a gRPC round‑trip.

 For older ScheduledWorkflows without the annotation, we still fall back to the API lookup to keep behavior compatible.

sduvvuri1603 committed:
feat: add pipeline run parallelism config

Introduce max_active_runs support end-to-end: compile semaphore settings,
propagate pipeline version metadata, and upsert parallelism ConfigMap entries
so Argo enforces per-version run limits. Expand/adjust backend, SDK, and test
coverage and include temporary debug logging to diagnose semaphore issues in
E2E parallelism tests.

Signed-off-by: sduvvuri1603 <sduvvuri@redhat.com>
Made-with: Cursor

hbelmiro committed:
fix: Improve parallelism E2E test robustness and fix ConfigMap race condition (#2)

* chore(tests): Refactor pipeline e2e tests to improve robustness and reduce flakiness

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

* chore(tests): Add E2eParallelism label and refactor pipeline parallelism e2e tests

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

* chore(backend): Use unique FieldManager per pipeline version to prevent key clobbering in shared ConfigMap

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

* chore(tests): Set default NUMBER_OF_NODES for E2EParallelism label in e2e-test workflow

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

---------

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

hbelmiro committed:
Bring back removed tests and other improvements (#3)

* chore(tests): Remove redundant limit clamping in pipeline concurrency e2e tests

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

* fix(backend): Reverted propagation policy removal

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

* fix(backend): Simplify `MaxActiveRuns` implementation and update handling of unset/invalid values

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

---------

Signed-off-by: Helber Belmiro <helber.belmiro@gmail.com>

hbelmiro reviewed (commented):
/lgtm

nsingla commented:
While trying to optimize the github copilot instructions, I ran this PR through it for a thorough review, here's what it gave me:
```
---                                                                                                                                                   
  PR #12442 Review: feat: Add pipeline run parallelism config
                                                                                                                                                        
  PR Overview                                               

  This PR adds max_active_runs to PipelineConfig, allowing pipeline versions to declare concurrent run limits using Argo semaphores backed by a
  ConfigMap. The flow is: SDK PipelineConfig → proto IR → Argo compiler (semaphore injection) → API server (ConfigMap upsert at run creation) → Argo
  enforcement.

  Scope: XXL PR (53 files, ~3,500 diff lines) crossing 7+ subsystems: Proto, SDK Python, Backend Go (apiserver, compiler, CRD controller, common/util),
  E2E tests, CI workflows, and Kustomize manifests.

  Per the impact-analysis.md: PRs >500 additions crossing 3+ subsystems should be flagged as "needs decomposition discussion." This PR would benefit
  from splitting CI/TLS changes (ensure-kfp-api.sh) into a separate PR since they are orthogonal to the parallelism feature.

  ---
  Critical Issues

  1. CRD controller imports apiserver/template — architectural violation
  backend/src/crd/controller/scheduledworkflow/controller.go:28

  import "github.com/kubeflow/pipelines/backend/src/apiserver/template"

  This violates the critical dependency direction rule: backend/src/crd/controller/* -> backend/src/crd/pkg only, never apiserver/. The controller
  should never depend on the API server package. This creates a coupling that makes independent deployment impossible and risks circular build
  dependencies.

  2. CRD controller makes synchronous API callback
  backend/src/crd/controller/scheduledworkflow/controller.go:1187-1193

  pipelineVersion, err := c.pipelineClient.GetPipelineVersion(ctx, &api.GetPipelineVersionRequest{...})

  The architecture guide explicitly states: "No synchronous API callbacks: Controllers must not call back to the API server. Pass data via CRD
  spec/annotations/labels." The extractMaxActiveRunsFromWorkflow method calls back to the API server to fetch the pipeline version spec, creating a
  circular dependency (API server → deploys controller → calls API server). If the API server is temporarily unavailable, scheduled workflow
  reconciliation will fail.

  Fix: The max_active_runs value should be passed via the ScheduledWorkflow CRD annotations/spec at creation time, not looked up at each reconciliation.
   The CreateJob / ScheduledWorkflow path in v2_template.go already sets the annotation — the controller should just read it from the workflow spec's
  synchronization semaphore configuration (which it partially does already) without needing to call back to the API server.

  3. Nested fire-and-forget goroutines in deletion path
  backend/src/common/util/workflow.go:877-905 and backend/src/apiserver/resource/resource_manager.go:513-531

  The deletion flow has nested goroutines:
  - DeletePipelineVersion → defer go func() → calls OnDeletePipelineVersion
  - OnDeletePipelineVersion → go func() → creates new K8s client → does cleanup

  The Go checklist explicitly flags: "No nested goroutines (go func() { go func() { ... } }())"

  Additional problems:
  - OnDeletePipelineVersion always returns nil because it launches a goroutine and returns immediately. The caller's error check is dead code.
  - The outer goroutine creates a 30s context, but OnDeletePipelineVersion creates its own 30s context internally — the outer context is effectively
  unused.
  - OnDeletePipelineVersion calls kubernetes.NewForConfig() inside the goroutine. The checklist says: "Reuse K8s clients from struct fields, not
  kubernetes.NewForConfig() per call."

  Fix: Remove the nested goroutine. Pass the K8s clientset from the WorkflowClient struct (or accept it as a parameter). Make the cleanup synchronous
  within a single goroutine with a single timeout context.

  ---
  Design Concerns

  4. extractMaxActiveRunsFromWorkflow is 115 lines with multiple fallback paths
  backend/src/crd/controller/scheduledworkflow/controller.go:675-1229

  This function has 3 different ways to find max_active_runs:
  1. Read from workflow synchronization semaphore
  2. Read from workflow annotations
  3. Call the API server to fetch the pipeline version spec, parse it as a template, and extract the value

  This complexity exists because the value isn't consistently passed through the system. If the annotation were always set at ScheduledWorkflow creation
   (as it is for one-off runs), this function would be ~10 lines.

  5. Behavioral change to CreateJob — recurring runs now resolve pipeline version eagerly
  backend/src/apiserver/resource/resource_manager.go:1175-1502

  When creating a recurring run with only PipelineId (no version), the code now fetches and stores the latest version's ID and name. The integration
  test confirms this with assert.Equal replacing assert.Empty for PipelineVersionID. This is a behavioral change that affects all recurring runs, not
  just those with parallelism configured. It should be explicitly documented in the PR description since it changes the semantics of "latest version"
  resolution from deferred (at run creation) to eager (at recurring run creation).

  6. ReplaceSDKInPipelineSpec global sleep modification
  backend/test/testutil/test_utils.go:149-152

  sleepDefaultPattern := regexp.MustCompile(`sleep_seconds:\s*int\s*=\s*20`)
  modifiedPipelineSpec = sleepDefaultPattern.ReplaceAllString(modifiedPipelineSpec, "sleep_seconds: int = 5")

  This modifies sleep durations in all pipeline specs globally, not just parallelism tests. This could cause unexpected behavior changes in unrelated
  tests. This modification should be scoped to only the parallelism pipeline files, or the parallelism pipeline files should use shorter sleeps
  natively.

  ---
  RBAC / Security

  7. RBAC changes look correct ✅
  - ConfigMap create is unscoped (required, since create can't use resourceNames)
  - get, update, patch are properly scoped to resourceNames: [kfp-argo-workflow-semaphores]
  - ClusterRoleBinding uses $(kfp-cluster-scoped-namespace) Kustomize variable ✅
  - New ClusterRole follows ml-pipeline-* naming convention ✅
  - Multi-user cluster role updated in sync ✅

  8. ConfigMap key validation
  backend/src/common/util/workflow.go:983-985

  Pipeline version IDs (UUIDs) are used as ConfigMap keys. While UUIDs match the [-._a-zA-Z0-9]+ pattern, there's no explicit validation. If a non-UUID
  value somehow reaches this path, it could create invalid ConfigMap keys. Consider adding a regex check.

  ---
  Test Coverage

  9. Missing unit tests for new Go functions
  Per the test quality guide: "Every new public function: happy path + edge cases + error paths." The following new functions lack unit tests:
  - GetRunNamespacesForPipelineVersion (run_store.go)
  - EnsurePipelineParallelismConfigMap (workflow.go)
  - deletePipelineParallelismConfigMapEntry (workflow.go)
  - OnDeletePipelineVersion (workflow.go)
  - applyPipelineParallelismConfigMapEntry (workflow.go)
  - extractMaxActiveRunsFromWorkflow (controller.go)

  For an XXL feature PR, the test quality guide requires: "Unit + integration + E2E + goldens + negative tests." The E2E and goldens are present, but
  dedicated unit tests for the backend logic are missing.

  10. validatePipelineRunSuccess now runs parallelism validation for all pipelines with max_active_runs
  backend/test/end2end/pipeline_e2e_test.go:824-835

  if limit, err := MaxActiveRuns(pipelineFilePath); err == nil {
      ValidateWorkflowParallelismAcrossRuns(...)
  }

  This adds parallelism validation as a side effect of the general validatePipelineRunSuccess function. This means any pipeline file that happens to
  have max_active_runs will trigger additional runs and validation, potentially adding significant runtime to non-parallelism test contexts. This should
   be explicitly called only from parallelism-specific test cases.

  ---
  Proto / SDK

  11. Proto change is backward-compatible ✅
  - New optional int32 max_active_runs = 3 — additive, optional, safe default (0 = no limit)
  - Go .pb.go regenerated
  - Binary test data files regenerated

  12. SDK PipelineConfig — missing docstring
  sdk/python/kfp/dsl/pipeline_config.py:106-107

  def __init__(self,
               workspace: Optional[WorkspaceConfig] = None,
               max_active_runs: Optional[int] = None):

  The max_active_runs parameter lacks a docstring. Per the Python checklist: "Docstrings on all public APIs/properties; include usage examples for new
  DSL options."

  ---
  Minor / Nits

  13. pipeline_with_string_machine_fields_task_output.yaml golden file has dependentTasks reordering. This appears to be a non-functional
  formatting-only change and should be noted in the PR description per the proto checklist: "Generated code formatting-only diffs noted in PR
  description."

  14. pipeline_with_workspace.yaml golden file has component hash changes and kfp version bumps (2.14.3 → 2.14.6). These are SDK version drift
  artifacts, not related to this PR's changes.

  15. ensure-kfp-api.sh is a well-written script with set -euo pipefail, proper error handling, and graceful fallback. However, it replaces
  forward-port.sh and adds TLS handling — this is orthogonal to the parallelism feature and should ideally be a separate PR.

  ---
  Summary

  ┌──────────────────────────────┬────────────────────────────────────────────────────┐
  │           Category           │                       Status                       │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ Proto backward compat        │ ✅ Pass                                            │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ SDK Python                   │ ⚠️ Missing docstring                               │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ Backend Go architecture      │ ❌ Controller→apiserver import violation           │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ Backend Go concurrency       │ ❌ Nested goroutines, per-call K8s client creation │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ RBAC / Security              │ ✅ Pass                                            │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ Tests proportional to change │ ⚠️ Missing unit tests for backend functions        │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ Golden files                 │ ✅ Pass (with unrelated drifts noted)              │
  ├──────────────────────────────┼────────────────────────────────────────────────────┤
  │ CI                           │ ⚠️ Orthogonal changes mixed in                     │
  └──────────────────────────────┴────────────────────────────────────────────────────┘

  Blocking: Issues #1, #2, and #3 should be resolved before merge. The CRD controller should not import apiserver packages or make synchronous API
  callbacks — the parallelism value should flow via annotations/CRD spec. The nested goroutine pattern should be simplified.
```

@sduvvuri1603 @hbelmiro @HumairAK , there seems to be few critical issues listed above
